### ISSUE TITLE
[Feature Request] Nintendo Switch Pro Controller Configuration Support

### ISSUE THREAD

**Issue by Vash63** [ValveSoftware/steam-for-linux#4920]
The Switch Pro Controller is natively recognized as X-Input and works under Bluetooth, however if I enable the new 'Generic Gamepad Configuration Support' feature my button mappings are gone and the controller no longer works in the Big Picture interface.

I'm not sure exactly what would be needed, but this is what I'm able to get out of udevadm:

```  looking at device '/devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.7/4-1.7:1.0/bluetooth/hci0/hci0:11/0005:057E:2009.001A/input/input48/event7':
    KERNEL=="event7"
    SUBSYSTEM=="input"
    DRIVER==""

  looking at parent device '/devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.7/4-1.7:1.0/bluetooth/hci0/hci0:11/0005:057E:2009.001A/input/input48':
    KERNELS=="input48"
    SUBSYSTEMS=="input"
    DRIVERS==""
    ATTRS{name}=="Pro Controller"
    ATTRS{phys}=="00:02:72:c6:98:e7"
    ATTRS{properties}=="0"
    ATTRS{uniq}=="98:b6:e9:40:e7:39"

  looking at parent device '/devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.7/4-1.7:1.0/bluetooth/hci0/hci0:11/0005:057E:2009.001A':
    KERNELS=="0005:057E:2009.001A"
    SUBSYSTEMS=="hid"
    DRIVERS=="hid-generic"
    ATTRS{country}=="21"
```

And just to clarify, the controller works fine in Big Picture with the legacy mapping system, this just does not work with games that are Xbox or PS controllers only (Such as Mad Max).

**Comment by Vash63** [comment]
Reviewed this in #steamlug on IRC also and was recommended to look at the SDL controllerdb file - added a bug there as well with the mappings generated by Steam:

https://bugzilla.libsdl.org/show_bug.cgi?id=3621

**Comment by Vash63** [comment]
Another update: libsdl has added the mappings I pulled from Steam, so this issue should resolve itself whenever the next SDL release happens and is pulled into Steam.

It may also be useful to keep this open as a feature request for full PS4-style support though, with motion control and proper LED support. The Switch is selling extremely well and could quickly be a popular controller for users.

**Comment by TruboXL** [comment]
As much as I hate the Xbox360 mapping, I feel weird having A,B,X,Y all inverted...

Is everyone ok with that?
Maybe I am the only one here arguing, huhu, here's my mapping instead (made via `antimicro`):
```
050000007e0500000920000001000000,Pro Controller,platform:Linux,a:b1,b:b0,x:b3,y:b2,back:b8,start:b9,guide:b12,leftshoulder:b4,rightshoulder:b5,leftstick:b10,rightstick:b11,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,righttrigger:b7,dpup:h0.1,dpleft:h0.8,dpdown:h0.4,dpright:h0.2,
```

**Comment by rockingdice** [comment]
Does it work now in latest Steam?

**Comment by TruboXL** [comment]
> Does it work now in latest Steam?

Yes, it works. Bluetooth and play. It follows Xbox 360 mappings so each button and stick is working, except the Capture button since 360 doesn't have an equivalent button. Since it follows 360 mapping, expect menus like Big Picture to work differently from a Nintendo fan thought it would.

Pro Controller's B for accept, go in (360's A)
Pro Controller's A for reject, go back (360's B)
X, Y are also inverted...
Steam seems to ignore "Guide button focuses Steam" so if you have Steam in the background and you are playing non-Steam game like Dolphin, you ended up causing Steam to go into foreground.

Like on Windows, there's no rumble, no plug and play through wired, no NFC and motion controls obviously. Player number LED still flashes left and right when connected.

Unlike on Windows, you can't calibrate the sticks easily on Linux, so something you need to think about.

**Comment by lordelan** [comment]
> Unlike on Windows, you can't calibrate the sticks easily on Linux, so something you need to think about.

That's something that still bothers me. In some games where you can "run" with the analog sticks from a 360, you can just "walk" with the ones from the Switch Pro Controller.
In Windows this can be solved by calibrating the controller ones but in Linux? Especially Ubuntu (which I'm using)?

**Comment by CodingKoopa** [comment]
Support for the Switch Pro Controller [has been added to the Steam Client Beta](https://steamcommunity.com/games/353370/announcements/detail/1648757912601251834), but I can't get Steam to recognize my controller.

I have my controller connected over bluetooth via the KDE Bluetooth System Settings Module provided by [Bluedevil](https://userbase.kde.org/Bluedevil):
![KDE Blutetooth System Settings Module screenshot, showing a connection to the Pro Controller.](https://user-images.githubusercontent.com/13039555/39555278-11a38cae-4e46-11e8-95d7-cc74c06c09ca.png)

In the [KDE Joystick System Settings Module](https://docs.kde.org/trunk5/en/kde-workspace/kcontrol/joystick/index.html), inputs from the controller are displayed:
![KDE Joystick System Settings Module screenshot, showing inputs on device /dev/input/js0.](https://user-images.githubusercontent.com/13039555/39555173-27b17b42-4e45-11e8-911d-610aa6b1ca41.png)

However, in the Steam controller settings, the Pro Controller is not detected:
![Steam controller settings screenshot, with no detected controllers.](https://user-images.githubusercontent.com/13039555/39555364-e07da578-4e46-11e8-83cc-ac60a07741af.png)

Before this update, I was able to use the Pro Controller in Rocket League, but currently neither Steam Big Picture nor Rocket League are detecting it.

**Comment by Phantop** [comment]
I'm having the same issue as CodingKoopa with my wireless controller. I'm using Solus Linux, so it isn't just a distro thing.

**Comment by Plagman** [comment]
Today's Beta build should fix the bug with analog sticks.

**Comment by CodingKoopa** [comment]
The new `May 3 2018, at 17:23:24` build fixed Bluetooth Pro Controller support for me, works perfectly now.

**Comment by Phantop** [comment]
The controller works but, in a partially unrelated issue, I can't get it to work with a game running under Wine. Steam refused to load it not matter what options I gave for a shortcut.

Is there any way to keep the controller in game mode always, so that other programs will recognize it as an Xbox controller?

**Comment by MrCapone** [comment]
Same problem here.

**Comment by Phantop** [comment]
I found out how to get desktop control to work. Go to desktop configuration and then the select template option. The regular gamepad option is still there.

**Comment by MrCapone** [comment]
When i connect pro controller by bluetooth it appear in jstest, but steam don't see it.
![2018-05-20 00-32-15](https://user-images.githubusercontent.com/5929042/40271289-6ce22314-5bc5-11e8-9d01-488b68d165e3.png)
![2018-05-20 00-32-35](https://user-images.githubusercontent.com/5929042/40271292-761964b0-5bc5-11e8-867b-14ac2a296565.png)
Maybe steam need special udev rule for pro controller?
My system Linux Mint 18.3 (Ubuntu 16.04)
Steam installed by .deb from official site version 10.0.0.54
steam version:
![2018-05-20 00-35-14](https://user-images.githubusercontent.com/5929042/40271319-de0a6bbe-5bc5-11e8-95fc-701bffbce992.png)

**Comment by Plagman** [comment]
Yes, can you try adding these udev rules?

```
# Nintendo Switch Pro Controller over USB hidraw
KERNEL=="hidraw*", ATTRS{idVendor}=="057e", ATTRS{idProduct}=="2009", MODE="0666"

# Nintendo Switch Pro Controller over bluetooth hidraw
KERNEL=="hidraw*", KERNELS=="*057E:2009*", MODE="0666"
```

**Comment by MrCapone** [comment]
@Plagman yes it works thanks. Tested usb and bluetooth.

**Comment by MrCapone** [comment]
@Plagman yes it works thanks. Tested usb and bluetooth.
UPD
But in some games left stick doesn't work :( For example stardew valley. But it works well in other games.

**Comment by yaomtc** [comment]
I'm unable to get the Switch Pro Controller working. I don't have any Bluetooth capability on this computer, so I've had to try using USB, but no matter what I do, Steam won't recognize it. I've tried installing steam-devices, I've tried disconnecting it from the Switch to avoid accidentally turning the Switch on, I've tried different USB ports... nothing works.

* Steam client version (build number or date): August 29 2018, package version 1535576546
* Distribution (e.g. Ubuntu): Ubuntu 18.04.1 (64 bit)
* Opted into Steam client beta?: [Yes/No] Yes
* Have you checked for system updates?: [Yes/No] Yes

```
Local Device Found
  type: 057e 2009
  path: 0003:0010:00
  serial_number:  - 0
  Manufacturer: 
  Product:      
  Release:      200
  Interface:    0

Unable to open local device: 0003:0010:00
```

[steam-procontroller-output.txt](https://github.com/ValveSoftware/steam-for-linux/files/2342597/steam-procontroller-output.txt)

[60-steam-input.rules.txt](https://github.com/ValveSoftware/steam-for-linux/files/2342587/60-steam-input.rules.txt)

**Comment by yaomtc** [comment]
Update: It works for me now after some update.
